### ISSUE TITLE
concurrent commands on different vagrants hangs on Windows/vmware

### ISSUE THREAD

**Issue by f3rrix** [hashicorp/vagrant#9447]
### Vagrant version
Vagrant 2.0.2
vagrant-vmware-workstation (5.0.4)

### Host operating system
Windows 7 Ultimate (x64)
Build 7601: Service Pack 1

### Guest operating system
Any

### Vagrantfile
IMO not relevant here; see below

### Debug output
First vagrant process hung: https://gist.github.com/f3rrix/43943bd46c94b9ece35aa9fc4a8d8860
Second vagrant process succeeded: https://gist.github.com/f3rrix/b4a99a361961ddc0637a7a1c78341c90
Third vagrant process hung: https://gist.github.com/f3rrix/8a8794cd8127d1c836060ff11f508f5c

### Expected behavior
Old-timey vagrant with vmware_workstation provider in the 1.x era could run simultaneously with impunity.  This allowed me to "up" several things at once and get them all booting, or perform an action on one system (e.g. reload) while another guest was provisioning.

### Actual behavior
Ever since (I think) version 2, I must be *scrupulously* careful to have a maximum of ONE vagrant command running anywhere on the system at once.  Otherwise some number of them will "hang" forever waiting for output from the sudo_helper scheduled task, even though it has already run and failed.

In my debug output test, I had three vagrants "racing" here.  The middle one succeeded all the way to completion, and the other two hung at various points.  

I have already tried changing the task settings to allow queuing or parallel running, but this didn't help and I assume that's because there's an argument/output file that lives in exactly one place.  

Can I disable this scheduled task loop and just have it run "the old way" directly launching the vmware commands under my (already administrator) context?  Or can you flock the argument file before launching the task until after it completes? 

Since the source for the vmware plugin is not available, I can't offer any other tangible suggestions.

### Steps to reproduce
1.  Prepare several command prompts, each in a separate directory containing a vagrant+vmware guest
2. "vagrant up" them all, as close to the same time as possible (it doesn't require too much accuracy to trip you up)
3. Notice that usually only one will "work", and the others will sit forever at some incomplete stage.
4. You will have one ruby.exe and vagrant.exe for each hung process, and it will hang after "DEBUG sudo_helper: Started VMware helper task. Waiting for result."

**Comment by f3rrix** [comment]
I learned some more through repeated trials, tracing things, watching what vagrant_vmware_desktop_sudo_helper_windows_amd64.exe does, and log outputs.

Sometimes the helper fails because it cannot find the correct output .tmp file.  This is because it has read the helper.args file, but it was written by another vagrant than the one we want.  (I can prove this by seeing that the helper's error output prints the *wrong* guid as the name of the file it's looking for.)

There seems to be a flock around "helper.args" that I assume was desired to prevent this, but it cannot work.  Browsing ruby's file.c, I think flock doesn't actually work cross-process in Windows.  It seems to be for controlling multi-threaded access to a file within *one* process (which is useless to vagrant).  I'm not 100% sure on that.  But:

By wrapping the whole ugly sudo_helper operation in a globally-named win32-mutex (NOT a ruby mutex, they aren't cross-process!), I can successfully serialize the parts that used to clobber each other.  It's incredibly hacky to do this without access to the legit source code for sudo_helper, but it would be trivial for someone at Hashicorp to implement this correctly.

Basically:
```
require 'win32/mutex'
resultVar = nil
Mutex.new(false, 'Global\vagrant-vmware-sudo-helper') do |m|
    m.wait
    # here you do all the things related to setting up arguments, running the process, waiting for results, deleting the args file
    #...blah blah, set resultVar to what you want
    m.release
end
resultVal
```
With mutual exclusion, parallel vagrants can once again run on the same system with the vmware provider.  Please fix this properly!

**Comment by sspringe** [comment]
We have the same problem. This behavior first occured with VMWare provider 4.0.25 where the task scheduler based VMWare interaction was introduced to get rid of the administrator requirement. 
For us this is a painful feature break because running vagrant parallel in larger test lab environments definitely makes sense as it boosts the throughput drastically.
Therefore we are also very interested in a proper fix of this behavior change!

**Comment by f3rrix** [comment]
I opened a ticket with Hashicorp and the tier-1 person said they were going to forward it to a developer.  No feedback since then (>1week) and the ticket auto-closed due to inactivity.

**Comment by sspringe** [comment]
I also had long discussions about the parallelization use case with the tier-1 person (however he did really good support job for us). After all it sounds to me that parallel operation is not considered as "normal" operation for Vagrant. Besides that I think that safe parallel operation can never be completely guaranteed due to limitations in the backend system (example: the initial "vagrant up" performs a clone operation on the VM which is technically a linked clone in VMWare. Linked clones cannot be safely created parallel, only full clones.). Moreover we found out that the download of a box must be strictly sequentialized as Vagrant doesn't take care of locking the local file system structures.
Anyway the VMWare provider 4.0.19  performs well in parallel use case with minor limitations and I hope they will return to that quality soon with the latest version of the VMWare provider.

**Comment by kikitux** [comment]
Sorry for the delay here.

As today, vmware/virtualbox doesn't work in parallel.

Docker does.

So this is not a bug.

in virtualbox, when we deploy from template, the box gets create to an intermediate box before renaming, so 2 boxes from the same template will collide.

For VMWare I may assume the same, however the task seems to be causing issues.

I will try to get a developer comment on this, and give a final answer on if making virtualbox/vmware in parallel would be something can be done or not.

**Comment by f3rrix** [comment]
> So this is not a bug

Even stipulating parallelization is problematic in vmware, that's an incorrect characterization of *this* problem.  VMware is never even involved here, the problem occurs between vagrant and the task scheduler, before any vmware processes are reached.

Your code tries to flock the arguments file for the scheduled task, but that does not actually work (ruby flock is ineffective across processes).  So, parallel vagrants confuse each other's scheduled task execution by clobbering the single arguments file, leading to this hanging issue.  

There is a SIMPLE FIX: add a Win32 mutex with a global namespace around these operations, so it serializes all vmware commands across processes.  I could do this in like 5 lines of code if I had access to your source file.  I already tested and proved this works, with a hacked workaround, I just can't provide you a patch to prove it because the ruby source is obfuscated.  I'd be happy to sign an NDA and fix this code for you.

In summary, it's OK that VMware cannot be called in parallel.  You have the power to easily make vagrant line up its calls and serialize them.  The open source community can't help you in this case due to (understandable) source code obfuscation for your licensed product.  So I'm appealing to you to have another look.  

Please let me know if any of the above is unclear.  I spent *a lot* of time internally figuring all this out so I could create a local workaround, and I'm quite confident I've characterized the problem correctly.

**Comment by chrisroberts** [comment]
Hi there,

This issue should be resolved with the latest VMware plugin release. The scheduled task is no longer used which removes the errors encountered when running processes in parallel. You can read more about the new plugin release here:

https://www.hashicorp.com/blog/introducing-the-vagrant-vmware-desktop-plugin

My apologies for the inconvenience this bug introduced.

Cheers!

**Comment by f3rrix** [comment]
Huzzah!
